### PR TITLE
Adding VPN

### PR DESCRIPTION
**Pull request by Nepomuceno**
## Description

Adding VPN to help us support 

## Type of Change

- [ ] Non-module change (e.g. CI/CD, documentation, etc.)
- [x] Azure Verified Module updates:
  - [ ] Bugfix containing backwards compatible bug fixes
    - [ ] Someone has opened a bug report issue, and I have included "Closes #{bug_report_issue_number}" in the PR description.
    - [ ] The bug was found by the module author, and no one has opened an issue to report it yet.
  - [x] Feature update backwards compatible feature updates.
  - [ ] Breaking changes.
  - [ ] Update to documentation

# Checklist

- [x] I'm sure there are no other open Pull Requests for the same update/change
- [x] My corresponding pipelines / checks run clean and green without any errors or warnings
- [x] I did run all  [pre-commit](https://azure.github.io/Azure-Verified-Modules/contributing/terraform/terraform-contribution-flow/#5-run-pre-commit-checks) checks

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -161,6 +161,28 @@ Type: `string`
 
 Default: `""`
 
+### <a name="input_create_vpn"></a> [create\_vpn](#input\_create\_vpn)
+
+Description: Create a VPN Gateway
+
+Type: `bool`
+
+Default: `true`
```
Comment (shanepeckham) [inline README.md:170]: This should be optional though right? It needs to be false by default and we should set it to true for our deployments.

Comment (Nepomuceno) [reply README.md:170]: perfect will set the defualt to false but I can put to true in the examples folder

File: examples/default/README.md [hunk 2]
```diff
@@ -67,6 +67,7 @@ module "test" {
   name                = random_string.name.id
   resource_group_name = azurerm_resource_group.this.name
   enable_telemetry    = var.enable_telemetry
+  create_vpn          = true
```
Comment (PabloZaiden) [inline examples/default/README.md:70]: Should the default example have both the vpn and jumpbox enabled?

File: modules/vpn/variables.tf [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+
+variable "tags" {
+  type        = map(string)
+  default     = null
+  description = "A map of tags to add to all resources"
+}
+
+variable "vpn_address_spaces" {
+  type        = list(string)
+  default     = ["192.168.0.0/24"]
```
Comment (PabloZaiden) [inline modules/vpn/variables.tf:73]: Wouldn't this (very common) default address space be likely to collide with the address space of the client's network?

Comment (Nepomuceno) [reply modules/vpn/variables.tf:73]: This si the default on the portal normally and on the documentation that is why I didput the same here. Hopefully if the customer has this he would not be creating this vpn

File: examples/default/README.md (hunk 2 above)
Comment (Nepomuceno) [reply examples/default/README.md:70]: I think for now we are using example as our testbed when we do have one of those 2 working correctly we can disable the other.

**Review by PabloZaiden — APPROVED** [review]

File: modules/vpn/_header.md [hunk 4]
```diff
@@ -0,0 +1 @@
+# Vpn Intenral Module for the patttern.
```
Comment (vyta) [inline modules/vpn/_header.md:1]:
```suggestion
# Internal VPN Module for the pattern
```

File: modules/vpn/main.tf [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+    aad_audience         = var.aad_audience
+    aad_issuer           = var.aad_issuer
+    aad_tenant           = var.aad_tenant
+    vpn_auth_types       = ["AAD"]
+    vpn_client_protocols = ["OpenVPN"]
+  }
+}
+
+
+resource "azurerm_private_dns_resolver" "base" {
```
Comment (vyta) [inline modules/vpn/main.tf:73]: I see the resolver but no rules, is the resolver required for VPN?

File: modules/vpn/main.tf [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+  location            = var.location
+  name                = var.vpn_dns_resolver_name
+  resource_group_name = var.resource_group_name
+  virtual_network_id  = var.vnet_id
+  tags                = var.tags
+}
+
+resource "azurerm_private_dns_resolver_inbound_endpoint" "base" {
```
Comment (vyta) [inline modules/vpn/main.tf:81]: If resolver is required, do we need output endpoint as well for dns resolver?

**Review by vyta — CHANGES_REQUESTED** [review]
Thanks for this @Nepomuceno! I have some questions regarding the dns resolver, and more generally making sure we can get the VPN solution working e2e before merging since you have not been able to access the workspace with VPN at the moment.

File: variables.tf [hunk 7]
```diff
@@ -193,10 +193,16 @@ variable "virtual_network_name" {
   description = "The name of the Virtual Network. If not provided, a name will be generated."
 }
 
-variable "vm_network_security_group_name" {
-  type        = string
-  default     = ""
-  description = "The name of the Network Security Group for the virtual machines subnet. If not provided, a name will be generated."
+variable "vm_network_security_group" {
```
Comment (vyta) [inline variables.tf:196]: I think we want to enforce NSG on the subnet (general security guidance), but make the rules configurable, no?

**Review by vyta — CHANGES_REQUESTED** [review]
Thanks @Nepomuceno, just a couple comments

**Review by vyta — APPROVED** [review]

**Pull request merged**